distributed/worker.py

```diff
@@ -2261,6 +2270,14 @@ def transition_resumed_missing(
         """
         return self._transition_from_resumed(ts, "missing", stimulus_id=stimulus_id)
 
+    def transition_resumed_released(self, ts: TaskState, *, stimulus_id: str):
+        if not ts.done:
+            ts.state = "cancelled"
+            ts._next = None
+            return {}, []
```

Comment:
What's going to eventually pick this up and move it out of cancelled if there are no recommendations and no next?

Reply:
TLDR Once `ts.done = True` is set, i.e. execute/flight is done


`ts._next` should never have been set for `cancelled`. When I implemented cancelled/resumed states I made a few mistakes. The only relevant _next_ state for cancelled is `released`. That's the entire point of the state. The worker was instructed to release a key but it can't because it is "stuck" waiting for _something_ to finish, i.e. either the execution thread or the gather_data coroutine.
Once execution/gather finishes, they'll recommend a transition, e.g. upon success they'll recommend a transition to memory. For example cancelled->memory will ensure the key is released.

Why is this logic not directly implemented as part of the gather_dep/execute result parser? Well, have a look at the code there. Particularly the gather_dep result parser/finally clause is the most frequent source of deadlocks because the logic just blows up.
There is a bit of a design philosophy behind this to break a big, complex decision up into many small decisions that can be made using _local context_ information.

Consider the following example

- T1 was instructed to be computed
- T1 is dispatched to the threadpool
- T1 is requested to be released
- T1 finishes

The result, i.e. once it finishes could be implemented as

```python
if result == "success":
    if ts.not_cancelled:
        put_key_in_memory()
    else:
        release_key()
else:
    result == "failed"
    
    if ts.not_cancelled:
        if ts.asked_to_be_fetched_instead:
            # Whether this is a valid thing for the scheduler to ask is out
            # of scope for this comment. It happens/happened
            reschedule_to_fetch_key()
        else:
            put_key_in_memory()
    else:
        release_key()
```

With this transition system, it instead becomes

```python
# executing result parser
# This only requires local context, the decision should be simple, straight forward
if result == "success":
    recommend_memory()
else:
    assert result == "success"
    recommend_error()

def transition_cancelled_error(...):
    assert stuff
    release_task()

def transition_cancelled_memory(...):
    assert stuff
    put_key_in_memory()
```

This decision tree is a bit more complex for gathering keys. I'm not 100% convinced anymore if this is the right approach but here we are right now. The recent refactoring will allow us mid-term to move away from this if we choose to do so.